internal/pkg/util/bin/bin.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+// Package bin provides access to system binaries
+package bin
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
+	"github.com/sylabs/singularity/internal/pkg/runtime/engines/config"
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+	singularity "github.com/sylabs/singularity/pkg/runtime/engines/singularity/config"
+)
+
+var cache struct {
+	sync.Once
+	cryptsetup string
+	err        error
+}
+
+// Cryptsetup looks for the "cryptsetup" program returning the absolute
+// path to it. If the cryptsetup program is not available, this function
+// returns a non-nil error.
+func Cryptsetup() (string, error) {
+	cache.Do(func() {
+		cfgpath := filepath.Join(buildcfg.SINGULARITY_CONFDIR, "singularity.conf")
+		cache.cryptsetup, cache.err = cryptsetup(cfgpath)
+		sylog.Debugf("Using cryptsetup at %q", cache.cryptsetup)
+	})
+
+	return cache.cryptsetup, cache.err
+}
+
+func cryptsetup(cfgpath string) (string, error) {
+	cfg := singularity.FileConfig{}
+	if err := config.Parser(cfgpath, &cfg); err != nil {
+		return "", errors.Wrap(err, "unable to parse singularity configuration file")
+	}
+
+	path := cfg.CryptsetupPath
+
+	if path == "" {
+		if buildcfg.CRYPTSETUP_PATH == "" {
+			return "", errors.New("unable to obtain path to cryptsetup program")
+		}
+
+		path = buildcfg.CRYPTSETUP_PATH
+	} else {
+		switch fi, err := os.Stat(path); {
+		case err != nil:
+			return "", errors.Wrapf(err, "unable to stat %s", path)
+
+		case fi.IsDir():
```

Comment:
Should we force one way or the other? Either store directory or the absolute path to the binary?

Reply:
I'm not happy with this but I'm following the example set by mksquashfs: in the configuration file the user can provide a directory or a filename. If it's a directory, it's the directory containing the program, just like this code does.